sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,539 @@
+      // we use Mongo splitVector command to get the split keys
+      BasicDBObject splitVectorCommand = new BasicDBObject();
+      splitVectorCommand.append("splitVector", database + "." + collection);
+      splitVectorCommand.append("keyPattern", new BasicDBObject().append("_id", 1));
+      splitVectorCommand.append("force", false);
+      // maxChunkSize is the Mongo partition size in MB
+      LOGGER.debug("Splitting in chunk of {} MB", desiredBundleSizeBytes / 1024 / 1024);
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
+      Document splitVectorCommandResult = mongoDatabase.runCommand(splitVectorCommand);
+      splitKeys = (List<Document>) splitVectorCommandResult.get("splitKeys");
+
+      List<BoundedSource<String>> sources = new ArrayList<>();
+      if (splitKeys.size() < 1) {
+        LOGGER.debug("Split keys is low, using an unique source");
+        sources.add(this);
+        return sources;
+      }
+
+      LOGGER.debug("Number of splits is {}", splitKeys.size());
+      for (String shardFilter : splitKeysToFilters(splitKeys, filter)) {
+        sources.add(this.withFilter(shardFilter));
+      }
+
+      return sources;
+    }
+
+    /**
+     * Transform a list of split keys as a list of filters containing corresponding range.
+     *
+     * <p>The list of split keys contains BSon Document basically containing for example:
+     * <ul>
+     *   <li>_id: 56</li>
+     *   <li>_id: 109</li>
+     *   <li>_id: 256</li>
+     * </ul>
+     * </p>
+     *
+     * This method will generate a list of range filters performing the following splits:
+     * <ul>
+     *   <li>from the beginning of the collection up to _id 56, so basically data with
+     *   _id lower than 56</li>
+     *   <li>from _id 57 up to _id 109</li>
+     *   <li>from _id 110 up to _id 256</li>
+     *   <li>from _id 257 up to the end of the collection, so basically data with _id greater
+     *   than 257</li>
+     * </ul>
+     *
+     * @param splitKeys The list of split keys.
+     * @param additionalFilter A custom (user) additional filter to append to the range filters.
+     * @return A list of filters containing the ranges.
+     */
+    private static List<String> splitKeysToFilters(List<Document> splitKeys, String
+        additionalFilter) {
+      ArrayList<String> filters = new ArrayList<>();
+      String lowestBound = null; // lower boundary (previous split in the iteration)
+      for (int i = 0; i < splitKeys.size(); i++) {
+        String splitKey = splitKeys.get(i).toString();
+        String rangeFilter = null;
+        if (i == 0) {
+          // this is the first split in the list, the filter defines
+          // the range from the beginning up to this split
+          rangeFilter = String.format("{ $and: [ {\"_id\":{$lte:Objectd(\"%s\")}}",
+              splitKey);
+        }
+        if (i == splitKeys.size() - 1) {
```

Comment:
You can rewrite these 3 mutually exclusive conditions to a slightly simpler form:
if (i == 0) {
...
} else if (i == splitKeys.size() - 1) {
...
} else {
...
}